File: Year 3/Parallel and Distributed Programming/ConvolutionMatrix Java/src/Convolutions/ConvolutionTask.java
package Convolutions;

import Singleton.Convolution;

public class ConvolutionTask extends Thread {
    private final int[][] inputMatrix;
    private final int[][] convolutionMatrix;
    private final int[][] resultMatrix;
    private final int start;
    private final int end;
    private final Convolution convolution = Convolution.getInstance();

    public ConvolutionTask(int[][] inputMatrix, int[][] convolutionMatrix, int[][] resultMatrix, int start, int end) {
        this.inputMatrix = inputMatrix;
        this.convolutionMatrix = convolutionMatrix;
        this.resultMatrix = resultMatrix;
        this.start = start;
        this.end = end;
    }

    @Override
    public void run() {
        int M = inputMatrix[0].length;
        for (int i = start; i < end; i++) {
            for (int j = 0; j < M; j++) {
                resultMatrix[i][j] = convolution.applyConvolution(inputMatrix, convolutionMatrix, i, j);
            }
        }
    }
}

File: Year 3/Parallel and Distributed Programming/Convolution Matrix/ConvolutionMatrix C++/ConvolutionMatrix C++ Parallel/Source Files/Convolutions/ConvolutionTask.cpp
#include "../../Header Files/Convolutions/ConvolutionTask.h"

ConvolutionTask::ConvolutionTask(const vector<vector<int>>& inputMatrix,
	const vector<vector<int>>& convolutionMatrix,
	vector<vector<int>>& resultMatrix,
	int start, int end) :
	inputMatrix(inputMatrix),
	convolutionMatrix(convolutionMatrix),
	resultMatrix(resultMatrix),
	start(start),
	end(end) {
	convolution = Convolution::getInstance();
}

void ConvolutionTask::run() {
	int M = inputMatrix[0].size();
	for (int i = start; i < end; ++i) {
		for (int j = 0; j < M; ++j) {
			resultMatrix[i][j] = convolution->applyConvolution(inputMatrix, convolutionMatrix, i, j);
		}
	}
}